File: avalon-modules/avalon-authorization/src/test/java/io/github/lishangbu/avalon/authorization/repository/UserRepositoryTest.java
package io.github.lishangbu.avalon.authorization.repository;

import io.github.lishangbu.avalon.authorization.entity.Role;
import io.github.lishangbu.avalon.authorization.entity.User;
import jakarta.annotation.Resource;
import java.util.HashSet;
import java.util.Optional;
import java.util.Set;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

/// 用户 Repository 测试
///
/// 覆盖用户的插入与按用户名查询逻辑，依赖 Liquibase 初始化的测试数据与关系
///
/// @author lishangbu
/// @since 2025/8/20
class UserRepositoryTest extends AbstractRepositoryTest {

  @Resource private UserRepository userRepository;
  @Resource private RoleRepository roleRepository;

  @Test
  void testInsert() {
    User user = new User();
    user.setPassword("{bcrypt}$2a$10$IlYJ6qn4gyXUL.CCLzlN4ujjzlfI.3UbB0VQrYSUmiaPKpcnxdU.G");
    user.setUsername("test2");
    roleRepository
        .findById(1L)
        .ifPresent(
            role -> {
              Set<Role> roles = new HashSet<>();
              roles.add(role);
              user.setRoles(roles);
            });
    userRepository.saveAndFlush(user);
    Optional<User> userOptional = userRepository.findUserWithRolesByUsername("test2");
    Assertions.assertTrue(userOptional.isPresent());
    User savedUser = userOptional.get();
    Assertions.assertEquals("test2", savedUser.getUsername());
    Assertions.assertTrue(savedUser.getPassword().startsWith("{bcrypt}"));
    Assertions.assertEquals(1, savedUser.getRoles().size());
  }

  @Test
  void testFindByUsername() {
    Optional<User> userOptional = userRepository.findUserWithRolesByUsername("admin");
    Assertions.assertTrue(userOptional.isPresent());
    User user = userOptional.get();
    Assertions.assertEquals("admin", user.getUsername());
    Assertions.assertTrue(user.getPassword().startsWith("{bcrypt}"));
    Assertions.assertEquals(2, user.getRoles().size());
  }
}

File: avalon-modules/avalon-authorization/src/test/kotlin/io/github/lishangbu/avalon/authorization/repository/UserRepositoryTest.kt
package io.github.lishangbu.avalon.authorization.repository

import io.github.lishangbu.avalon.authorization.entity.User
import io.github.lishangbu.avalon.authorization.entity.addBy
import io.github.lishangbu.avalon.jimmer.support.readOrNull
import jakarta.annotation.Resource
import org.babyfish.jimmer.sql.ast.mutation.SaveMode
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

/**
 * 用户 Repository 测试
 *
 * 覆盖用户的插入与按用户名查询逻辑，依赖数据库初始化的测试数据与关系
 *
 * @author lishangbu
 * @since 2025/8/20
 */
class UserRepositoryTest : AbstractRepositoryTest() {
    @Resource
    private lateinit var userRepository: UserRepository

    @Resource
    private lateinit var roleRepository: RoleRepository

    @Test
    fun testInsert() {
        val role = requireNotNull(roleRepository.findNullable(1L))
        val user =
            userRepository.save(
                User {
                    hashedPassword =
                        "{bcrypt}\$2a\$10\$IlYJ6qn4gyXUL.CCLzlN4ujjzlfI.3UbB0VQrYSUmiaPKpcnxdU.G"
                    username = "test2"
                    phone = "13800000001"
                    email = "test2@example.com"
                    avatar = "https://example.com/avatar/test2.png"
                    roles().addBy(role)
                },
                SaveMode.INSERT_ONLY,
            )
        assertNotNull(user.id)

        val savedUser = requireNotNull(userRepository.loadByAccountWithRoles("13800000001"))
        assertEquals("test2", savedUser.username)
        assertEquals("https://example.com/avatar/test2.png", savedUser.avatar)
        assertTrue(savedUser.hashedPassword!!.startsWith("{bcrypt}"))
        assertEquals(1, savedUser.roles.size)
    }

    @Test
    fun testFindByIdDoesNotFetchRolesByDefault() {
        val user = requireNotNull(userRepository.findNullable(1L))
        assertNull(user.readOrNull { roles })
    }

    @Test
    fun testFindByIdWithRoles() {
        val user = requireNotNull(userRepository.findNullable(1L, AuthorizationFetchers.USER_WITH_ROLES))
        assertEquals(2, user.roles.size)
    }

    @Test
    fun testFindByAccountWithRoles() {
        val user = requireNotNull(userRepository.loadByAccountWithRoles("admin"))
        assertEquals("admin", user.username)
        assertNull(user.avatar)
        assertTrue(user.hashedPassword!!.startsWith("{bcrypt}"))
        assertEquals(2, user.roles.size)
    }
}
